engine/src/main/java/org/pentaho/di/trans/steps/orabulkloader/OraBulkDataOutput.java

```diff
@@ -0,0 +1,254 @@
+
+/**
+ * Does the opening of the output "stream". It's either a file or inter process communication which is transparant to
+ * users of this class.
+ *
+ * @author Sven Boden
+ * @since 20-feb-2007
+ */
+public class OraBulkDataOutput {
+  private OraBulkLoaderMeta meta;
+  private Writer output = null;
+  private StringBuilder outbuf = null;
+  private boolean first = true;
+  private int[] fieldNumbers = null;
+  private String enclosure = null;
+  private SimpleDateFormat sdfDate = null;
+  private SimpleDateFormat sdfDateTime = null;
+  private String recTerm = null;
+
+  public OraBulkDataOutput( OraBulkLoaderMeta meta, String recTerm ) {
+    this.meta = meta;
+    this.recTerm = recTerm;
+  }
+
+  public void open( VariableSpace space, Process sqlldrProcess ) throws KettleException {
+    String loadMethod = meta.getLoadMethod();
+    try {
+      OutputStream os;
+
+      if ( OraBulkLoaderMeta.METHOD_AUTO_CONCURRENT.equals( loadMethod ) ) {
+        os = sqlldrProcess.getOutputStream();
+      } else {
+        // Else open the data file filled in.
+        String dataFilePath = getFilename( getFileObject( space.environmentSubstitute( meta.getDataFile() ), space ) );
+        File dataFile = new File( dataFilePath );
+        // Make sure the parent directory exists
+        dataFile.getParentFile().mkdirs();
+        os = new FileOutputStream( dataFile, false );
+      }
+
+      String encoding = meta.getEncoding();
+      if ( Utils.isEmpty( encoding ) ) {
+        // Use the default encoding.
+        output = new BufferedWriter( new OutputStreamWriter( os ) );
+      } else {
+        // Use the specified encoding
+        output = new BufferedWriter( new OutputStreamWriter( os, encoding ) );
+      }
+    } catch ( IOException e ) {
+      throw new KettleException( "IO exception occured: " + e.getMessage(), e );
+    }
+  }
+
+  public void close() throws IOException {
+    if ( output != null ) {
+      output.close();
+    }
+  }
+
+  Writer getOutput() {
+    return output;
+  }
+
+  private String createEscapedString( String orig, String enclosure ) {
+    StringBuilder buf = new StringBuilder( orig );
+
+    Const.repl( buf, enclosure, enclosure + enclosure );
+    return buf.toString();
+  }
+
+  @SuppressWarnings( "ArrayToString" )
+  public void writeLine( RowMetaInterface mi, Object[] row ) throws KettleException {
+    if ( first ) {
+      first = false;
+
+      enclosure = meta.getEnclosure();
+
+      // Setup up the fields we need to take for each of the rows
+      // as this speeds up processing.
+      fieldNumbers = new int[meta.getFieldStream().length];
+      for ( int i = 0; i < fieldNumbers.length; i++ ) {
+        fieldNumbers[i] = mi.indexOfValue( meta.getFieldStream()[i] );
+        if ( fieldNumbers[i] < 0 ) {
+          throw new KettleException( "Could not find field " + meta.getFieldStream()[i] + " in stream" );
+        }
+      }
+
+      sdfDate = new SimpleDateFormat( "yyyy-MM-dd" );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The SimpleDateFormat object was created without setting the time zone. Make sure that you want to use the default time zone or use letter z, Z or X in the pattern. Otherwise, if you do not call setTimeZone() on the created SimpleDateFormat object, you might get an unexpected date and time when using the object.